### DISCUSSION TITLE
Let `rmi -af` remove pods as well

### DISCUSSION THREAD

**Discussion by vrothberg** [containers/podman#12607] [Ideas]
`podman rmi -af` is able to remove containers but is NOT able to remove pods which I think is an inconsistency.  If all containers would be removed, we can nuke the pods' infra containers as well.

@containers/podman-maintainers PTAL. If there's consensus I would like to change that behavior before 4.0.

```
~ $ podman rmi -af
Untagged: docker.io/library/alpine:latest
Deleted: c059bfaa849c4d8e4aecaeb3a10c2d9b3d85f5165c66ad3a4d937758128c4d18
Error: error removing image ed210e3e4a5bae1237f1bb44d72a05a2f1e5c6bfe7a7e73da179e2534269c459: container 8b32ea2a4c2fc8972d694b87c10ac4df7bd6e7a0ffec0aec8b32e10bdf299c05 using image could not be removed: container 8b32ea2a4c2fc8972d694b87c10ac4df7bd6e7a0ffec0aec8b32e10bdf299c05 is the infra container of pod 6364bc5b0b40d7b9cff4d4e6b337b29cff83d207d41dc1e966512b6aef53b8f9 and cannot be removed without removing the pod
```

**Comment by Luap99** [comment]
The same applies to `podman rm -af`

**Reply by vrothberg** [reply]
I didn't notice that yet but think it should be changed as well.

**Reply by mheon** [reply]
The problem is really with the API - `rm` suddenly needs to be able to report that it removed multiple containers, not just one, if we call via the remote API with `force`

**Reply by vrothberg** [reply]
I think that's something we can handle on a higher abstraction level (i.e., `pkg/domain/infra`).

**Reply by mheon** [reply]
Is it? If I'm a consumer of the remote API via Python and my single call to delete a container removes 5 containers and the pod they're a part of, that seems like something the API should inform me of. The other APIs aren't specific to containers, so the potential for removing >1 container is always there (more than one container could be using a given image, or network), but container rm has always removed only a single container, without side effects.

**Reply by vrothberg** [reply]
I totally agree and think the remote API should remain untouched _but_ we can handle it on the client side in `pkg/domain/infra/{abi,tunnel}`: 
```golang
if ctr.IsInfra() { RemovePod(...) }
```

**Reply by mheon** [reply]
Oh, I see.

I was actually advocating for the remote API to be changes to return an array of removed containers, not just one, so we can actually remove the entire pod but report that we did so - but this also works.

**Reply by vrothberg** [reply]
Since the remote API returns an array now, I will take a stab :hocho:

**Reply by vrothberg** [reply]
https://github.com/containers/podman/pull/12826/files

**Comment by Luap99** [comment]
Also `podman network rm -f netname` will remove the pod and does not error, see: https://github.com/containers/podman/blob/b01a421f3413ba01b2c189b82c8153bdbd2a05fb/pkg/domain/infra/abi/network.go#L88-L99

**Reply by vrothberg** [reply]
Excellent sales pitch! Changing `rmi`and `rm`would make the commands consistent

**Comment by TomSweeneyRedHat** [comment]
WDDD?  (what does Docker do?)

**Reply by giuseppe** [reply]
I guess they don't have this problem since there is no concept of "pod"

**Reply by vrothberg** [reply]
I concur. Docker doesn't support pods, so they don't need to to worry about them.

**Reply by rhatdan** [reply]
I agree and I think this is a good idea.  --force means --force.

**Reply by vrothberg** [reply]
Great. I'll take a stab at it.
